### PR TITLE
internal/function: use cache for expensive language operations 

### LINKED ISSUE #442: Add a intermediate cache for language UDF
language UDF can be very slow when querying a lot of files. Sometimes this is executing language for the same file multiple times. For example, if we iterate the `files` table the same file is multiplied by the number of commits.

It could be good to add an intermediate cache so the results from previous run can be used instead of finding again the language. We can use an LRU using the file name and blob hashes. We may be able to use the hash of the first bytes of the blob that enry uses to speed up the process.

### PR DESCRIPTION
**Pull request by erizocosmico**
Closes #442 

This adds a two queue lru cache to language UDF. Why a 2Q cache instead of a regular LRU? Because we do not only want the most recent but the most frequent as well to maximize the cache usage.

Since running enry on a file_path only is just marginally slower than using the cache, I decided to not use the cache unless the blob_content is being provided (we could use two caches, though) to not pollute the cache and use it only for the really expensive operations: reading stuff from blobs.

Note that the cache is per gitbase instance.

Results are pretty much the ones @ajnavarro describes in the issue.

Before:

```
mysql> select count(*) from (select file_path, language(file_path, blob_content) as language FROM files WHERE language(file_path, blob_content) = 'Go') t;
+----------+
| COUNT(*) |
+----------+
|    39781 |
+----------+
1 row in set (20.85 sec)
```

Now:

```
mysql> select count(*) from (select file_path, language(file_path, blob_content) as language FROM files WHERE language(file_path, blob_content) = 'Go') t;
+----------+
| COUNT(*) |
+----------+
|    39781 |
+----------+
1 row in set (17.14 sec)
```

### CODE REVIEW

File: internal/function/language.go [hunk 1]
```diff
@@ -1,12 +1,27 @@
...
+	"encoding/binary"
 	"fmt"
+	"hash/crc32"
 
+	lru "github.com/hashicorp/golang-lru"
 	enry "gopkg.in/src-d/enry.v1"
 	"gopkg.in/src-d/go-mysql-server.v0/sql"
 )
 
+const defaultLanguageCacheSize = 10000
```
Comment (kuba--) [inline internal/function/language.go]: Right now, we are trying to let apps adjust cache sizes (and pass them down to packages like go-git), so maybe in this case we can also think about how easily we can modify this cache in the future (at least size).

Comment (erizocosmico) [reply internal/function/language.go]: Yep, this should be configurable, right @ajnavarro? Environment variable or CLI flag that passes down to here? WDYT?

Comment (ajnavarro) [reply internal/function/language.go]: yep, it should be configurable. it should be a env var o cli flag, yep.

**Comment by erizocosmico** [comment]
Added configuration to the cache size via env var

File: docs/using-gitbase/configuration.md [hunk 2]
```diff
@@ -12,6 +12,7 @@
 | `GITBASE_INDEX_DIR`          | directory to save indexes, default `/var/lib/gitbase/index`                        |
 | `GITBASE_TRACE`              | enable jaeger tracing, default disabled                                            |
 | `GITBASE_READONLY`           | allow read queries only, disabling creating and deleting indexes, default disabled |
+| `GITBASE_LANGUAGE_CACHE_SIZE`           | size of the cache for the `language` UDF. The size is the maximum number of elements kept in the cache |
```
Comment (ajnavarro) [inline docs/using-gitbase/configuration.md]: shall we add the default value here?

File: internal/function/language.go [hunk 3]
```diff
@@ -1,12 +1,42 @@
...
+	"os"
+	"strconv"
 
+	lru "github.com/hashicorp/golang-lru"
 	enry "gopkg.in/src-d/enry.v1"
 	"gopkg.in/src-d/go-mysql-server.v0/sql"
 )
 
+const (
+	languageCacheSizeKey     = "GITBASE_LANGUAGE_CACHE_SIZE"
```
Comment (ajnavarro) [inline internal/function/language.go]: should we add that env variable as we are doing with others? check https://github.com/src-d/gitbase/blob/master/cmd/gitbase/command/server.go#L50

Comment (erizocosmico) [reply internal/function/language.go]: If we add it there we will have to pass it all the way down until the infinite. Besides, there are other examples of environment variables such as blob max size that are not there but where they're used

**Comment by erizocosmico** [comment]
Added default value to docs

**Pull request merged**